YOLO Object Detection Example (#1795)
ROBOTIS-GIT/emanual: _includes/en/platform/turtlebot3/basic_examples/basic_examples_yolo_object_detection_humble.md

**Step 1: Customize the Script**  
If you haven't already cloned and built the `turtlebot3_applications` package, run thr following commands first.  
**[TurtleBot3 SBC]**  
```bash
$ cd ~/turtlebot3_ws/src/
$ git clone -b humble https://github.com/ROBOTIS-GIT/turtlebot3_applications.git
$ cd ~/turtlebot3_ws && colcon build --symlink-install
``` 

Open and edit the following file to update the model path and other parameters.  
**[TurtleBot3 SBC]**  
```bash
$ nano ~/turtlebot3_ws/src/turtlebot3_applications/turtlebot3_yolo_object_detection/turtlebot3_yolo_object_detection/turtlebot3_yolo_object_detection.py
``` 

Update the line with the correct path to your `best.pt` file.
```bash
self.model = YOLO("/home/ubuntu/best.pt")  # Update with your actual model path
``` 

Then, build your workspace.  
**[TurtleBot3 SBC]**  
```bash
$ cd ~/turtlebot3_ws
$ colcon build --packages-select turtlebot3_yolo_object_detection
``` 

**Step 2: Launch the Camera Node**  
**[TurtleBot3 SBC]**  
```bash
 $ ros2 launch turtlebot3_bringup camera.launch.py
 ``` 

**Step 3: Run the Detection Node**  
**[TurtleBot3 SBC]**  
```bash

> $ ros2 run turtlebot3_example turtlebot3_object_detection_node

Author: [513e3c5] Fixed

**What is YOLO?**  
YOLO(You Only Look Once) is a real-time object detection model. It views the entire image in a single pass("only once") and predicts both the bounding boxes and class probabilities directly. This makes YOLO extremely fast and efficient, making it ideal for real-time applications on TurtleBot3.  

> > In this example, we use *YOLOv8* to perform real-time object detection using the TurtleBot3 camera. We will train a custom model and integrate it into a ROS2 node for deployment on the robot. By the end of this tutorial, you’ll be able to train a YOLO model to recognize specific objects and run real-time detection directly on your TurtleBot3.   

Reviewer: ```suggestion
> In this example, we use *YOLOv8* to perform real-time object detection using the TurtleBot3 camera. We will train a custom model and integrate it into a ROS 2 node for deployment on the robot. By the end of this tutorial, you’ll be able to train a YOLO model to recognize specific objects and run real-time detection directly on your TurtleBot3.   
```

- Example of `data.yaml` file  
![](/assets/images/platform/turtlebot3/basic_examples/yolo_object_detection/yaml.png) 

**Step 4: Train the Model**  
**[Colab]**  
```bash
from ultralytics import YOLO
model=YOLO("yolov8n.pt")
model.train(data='{data_path}/data.yaml',epochs=100)
```  
- In this example, we use `yolov8n.pt`, a lightweight model optimized for speed and efficiency. You may use other variants like `yolov8s`, `m`, or `l` based on your needs.  
- `epochs` refers to the number of times the model iterates over the entire dataset during training. If set too low, the model may not learn effectively. If set too high, training can take a long time or exceed Google Colab's GPU usage limit.  
- `data_path` refers to the directory where your dataset and `data.yaml` file are located.  

**Step 5: Download the Trained Model**  
After training, the best-performing weights(`best.pt`) will be saved at `/content/runs/detect/train/weights`.  
**[Colab]**  
```bash
from google.colab import files
files.download('/content/runs/detect/train/weights/best.pt')
```  

### [**Remote PC Setup**](#remote-pc-setup)  

> **Step 1: ROS2 Package Setup**  

Reviewer: ```suggestion
**Step 1: ROS 2 Package Setup**  
```

Open and edit the following file to update the model path and other parameters.  
**[Remote PC]**  
```bash
$ nano ~/turtlebot3_ws/src/turtlebot3_applications/turtlebot3_yolo_object_detection/turtlebot3_yolo_object_detection/turtlebot3_yolo_object_detection.py
``` 
Update the line with the correct path to your `best.pt` file.
```bash
self.model = YOLO("~/Downloads/best.pt")  # Update with your actual model path
``` 
Then, build your workspace.  
**[Remote PC]**  
```bash
$ cd ~/turtlebot3_ws && colcon build --symlink-install --packages-select turtlebot3_yolo_object_detection
$ source install/setup.bash
``` 

**Step 3: Install the Required Dependencies**  
Install PyTorch and Ultralytics.  
Visit the official [PyTorch Installation Guide](https://pytorch.org/get-started/locally/) to install the correct version for your system.  
**[Remote PC]**  
```bash
# For CPU-only (example)
$ pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cpu
```  
Once PyTorch is installed,  
**[Remote PC]**  
```bash
$ pip3 install ultralytics opencv-python "numpy<2.0"
```  

### [**Camera Stream Setup**](#camera-stream-setup)  

**Launch the Camera Node**  

> Ensure that the SBC and Remote PC are on the same network and ROS2 DDS communication is properly set(e.g., `ROS_DOMAIN_ID`, `ROS_LOCALHOST_ONLY=0`).  

Reviewer: ```suggestion
Ensure that the SBC and Remote PC are on the same network and ROS 2 DDS communication is properly set(e.g., `ROS_DOMAIN_ID`, `ROS_LOCALHOST_ONLY=0`).  
```

Author: [a47e838] Fixed
